File: 783.minimum-distance-between-bst-nodes.cs
/*
 * @lc app=leetcode id=783 lang=csharp
 *
 * [783] Minimum Distance Between BST Nodes
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution 
{
    int minDiff = Int32.MaxValue / 2;
    int lastVal = Int32.MinValue / 2;
    public void dfs(TreeNode node)
    {
        if (node.left is not null)
        {
            dfs(node.left);
        }
        minDiff = Math.Min(minDiff, node.val - lastVal);
        lastVal = node.val;
        if (node.right is not null)
        {
            dfs(node.right);
        }
    }
    public int MinDiffInBST(TreeNode root) 
    {
        dfs(root);
        return minDiff;
    }
}
// @lc code=end

File: 783.minimum-distance-between-bst-nodes.py
#
# @lc app=leetcode id=783 lang=python3
#
# [783] Minimum Distance Between BST Nodes
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Optional


class Solution:
    def dfs(self,root):
        if root.left:
            self.dfs(root.left)
        self.min_diff = min(self.min_diff, root.val -  self.last_node_val)
        self.last_node_val = root.val
        if root.right:
            self.dfs(root.right)
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        self.min_diff = float('inf')
        self.last_node_val = -float('inf')
        self.dfs(root)
        return self.min_diff
# @lc code=end

